lib/datalog/access.go

```diff
@@ -0,0 +1,380 @@
+	r.createRoleMapping(role)
+	for _, login := range role.GetLogins(types.Allow) {
+		r.addPredicate(roleAllowsLogin, role.GetName(), login)
+	}
+	for _, login := range role.GetLogins(types.Deny) {
+		r.addPredicate(roleDeniesLogin, role.GetName(), login)
+	}
+	for key, values := range role.GetNodeLabels(types.Allow) {
+		for _, value := range values {
+			r.addPredicate(roleAllowsNodeLabel, role.GetName(), key, value)
+		}
+	}
+	for key, values := range role.GetNodeLabels(types.Deny) {
+		for _, value := range values {
+			r.addPredicate(roleDeniesNodeLabel, role.GetName(), key, value)
+		}
+	}
+}
+
+func (r *AccessResponse) createNodePredicates(node types.Server) {
+	r.createNodeMapping(node)
+	for key, value := range node.GetAllLabels() {
+		r.addPredicate(nodeHasLabel, node.GetHostname(), key, value)
+	}
+	r.addPredicate(nodeHasLabel, node.GetHostname(), types.Wildcard, types.Wildcard)
+}
+
+func (r *AccessResponse) generateAtomStrings(key string) [][]string {
+	var ret [][]string
+	for _, pred := range r.Accesses[key] {
+		var atoms []string
+		for _, atom := range pred.Atoms {
+			atoms = append(atoms, r.reverseMappings[atom])
+		}
+		ret = append(ret, atoms)
+	}
+	return ret
+}
+
+// BuildStringOutput creates the UI for displaying access responses.
+func (r *AccessResponse) BuildStringOutput() string {
+	accessTable := asciitable.MakeTable([]string{"User", "Login", "Node", "Allowing Roles"})
+	allowingRoles := make(map[string][]string)
+	accessStrings := r.generateAtomStrings(accesses)
+	for _, atoms := range accessStrings {
+		key := createDupMapKey(atoms[:roleIndex])
+		allowingRoles[key] = append(
+			allowingRoles[key],
+			atoms[roleIndex],
+		)
+	}
+	for _, key := range sortKeys(allowingRoles) {
+		sort.Strings(allowingRoles[key])
+		accessTable.AddRow(append(strings.Split(key, keyJoin), strings.Join(allowingRoles[key], ", ")))
+	}
+
+	denyTable := asciitable.MakeTable([]string{"User", "Logins", "Node", "Denying Role"})
+	deniedLogins := make(map[string][]string)
+	denyStrings := r.generateAtomStrings(denyAccesses)
+	for _, atoms := range r.generateAtomStrings(denyLogins) {
+		atomList := append(atoms[:nodeIndex], atoms[loginIndex:]...)
+		atomList[nodeIndex] = types.Wildcard
+		denyStrings = append(denyStrings, atomList)
+	}
+	for _, atoms := range denyStrings {
+		key := createDupMapKey(remove(atoms, loginIndex))
+		deniedLogins[key] = append(
+			deniedLogins[key],
+			atoms[loginIndex],
+		)
+	}
+	for _, key := range sortKeys(deniedLogins) {
+		atomList := strings.Split(key, keyJoin)
+		atomList = append(atomList[:nodeIndex], atomList[loginIndex:]...)
+		sort.Strings(deniedLogins[key])
+		atomList[loginIndex] = strings.Join(deniedLogins[key], ", ")
+		denyTable.AddRow(atomList)
+	}
+
+	denyOutputString := generateTableString(len(denyStrings) == 0 && len(r.Accesses[denyLogins]) == 0, denyTable, denyNullString)
+	accessOutputString := generateTableString(len(accessStrings) == 0, accessTable, accessNullString)
```

Comment:
This is a bit hard to read - instead of passing condition to the function, just do plain "if". It will be much clearer to the reader and you won't need the "generateTableString" function (and the "xxxNullString" variables either).